# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**


- 0–100000 — `'D'`;
- 100001–150000 — `'C'`;
- 150001–240000 — `'B'`;
- 240001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 100000:
            return 'D'
        elif 100001 <= income <= 150000:
            return 'C'
        elif 150001 <= income <= 240000:
            return 'B'
        elif income >= 240001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

*План работы*

Чтобы определить зависимость между количеством детей и возвратом кредита в срок, сгруппируем данные по количеству детей из столбца `children` и посчитаем конверсию возврата кредита в срок `debt` для каждой группы. Чем меньше будет конверсия, тем меньше должников в определенной категории, а следовательно и больше вероятность возврата кредита в срок.

In [22]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# сгруппируем данные data по количеству детей children 
# и применим метод agg(), чтобы для каждой группы посчитать sum() и count() 
children_grouped = data.groupby('children').agg({'debt': ['sum', 'count']})
# посчитаем конверсию для каждой группы и создадим новый столбец conversion, для этого нужно обратиться 
# в столбце debt к каждому подстолбцу sum и count
children_grouped['conversion'] = children_grouped['debt']['sum'] / children_grouped['debt']['count']


In [23]:
# найдем долю заемщиков в каждой категории от общего количества заемщиков
children_grouped['category_part'] = children_grouped['debt']['count'] / children_grouped['debt']['count'].sum()
children_grouped

debt        conversion category_part
           sum  count                         
children                                      
0         1063  14091   0.075438      0.660588
1          444   4808   0.092346      0.225400
2          194   2052   0.094542      0.096198
3           27    330   0.081818      0.015470
4            4     41   0.097561      0.001922
5            0      9   0.000000      0.000422

**Вывод: самая низкая конверсия для семей с пятью детьми, но таких семей очень мало, и их данные не стоит брать за основу. Если убрать из внимания семей из пяти детей, то значения конверсий примерно одинаковы, поэтому я бы не рекомендовала банкам принимать решение о выдаче кредита на основе количества детей, учитывая еще тот факт, что по этим данным, семьи без детей берут кредиты чаще, чем семьи с детьми.** 

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [24]:
# Ваш код будет здесь. Вы можете создавать новые ячейки
# сгруппируем данные data по семейному положению family_status 
# и применим метод agg(), чтобы для каждой группы посчитать sum() и count() 
family_status_grouped = data.groupby('family_status').agg({'debt': ['sum', 'count']})
family_status_grouped['conversion'] = family_status_grouped['debt']['sum'] / family_status_grouped['debt']['count']


In [25]:
# найдем долю заемщиков в каждой категории от общего количества заемщиков
family_status_grouped['category_part'] = family_status_grouped['debt']['count'] / family_status_grouped['debt']['count'].sum()
family_status_grouped

debt        conversion category_part
                       sum  count                         
family_status                                             
Не женат / не замужем  273   2796   0.097639      0.131077
в разводе               84   1189   0.070648      0.055740
вдовец / вдова          63    951   0.066246      0.044583
гражданский брак       385   4134   0.093130      0.193802
женат / замужем        927  12261   0.075606      0.574797

**Вывод: больше половины заемщиков состоят в браке и конверсия у них ниже, чем у тех категорий граждан, которые не состоят в официальном браке (не женат / не замужем, гражданский брак), самая низкая конверсия у тех, кто в разводе, а также у вдовцов / вдов, но их суммарно всего 10% от общего количества заемщиков, поэтому я бы рекомендовала банкам обращать внимание на семейное положение заемщиков.** 

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [26]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# сгруппируем данные data по категориям доходов total_income_category 
# и применим метод agg(), чтобы для каждой группы посчитать sum() и count() 
total_income_category_grouped = data.groupby('total_income_category').agg({'debt': ['sum', 'count']})
total_income_category_grouped['conversion'] = total_income_category_grouped['debt']['sum'] / \
total_income_category_grouped['debt']['count']

In [27]:
# найдем долю заемщиков в каждой категории от общего количества заемщиков
total_income_category_grouped['category_part'] = total_income_category_grouped['debt']['count'] / \
total_income_category_grouped['debt']['count'].sum()


Добавим столбец `total_income_amount_rub`, чтобы понимать, какая сумма доходов в рублях относится к определенной категории.

In [28]:
total_income_category_grouped['total_income_amount_rub'] = \
['240001 и выше', '150001–240000', '100001–150000', '0–100000']
total_income_category_grouped

debt       conversion category_part  \
                       sum count                            
total_income_category                                       
A                      217  3113   0.069708      0.145938   
B                      542  6664   0.081333      0.312409   
C                      619  7110   0.087060      0.333318   
D                      354  4444   0.079658      0.208335   

                      total_income_amount_rub  
                                               
total_income_category                          
A                               240001 и выше  
B                               150001–240000  
C                               100001–150000  
D                                    0–100000


**Существует зависимость между уровнем дохода и возвратом кредита в срок. Данные показали, чем выше доходы у заемщика, тем большая вероятность возврата кредита в срок, т.к. конверсия для категории заемщиков с наибольшими доходами наименьшая, а, значит, и количество должников в этой группе меньше**


#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [29]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# сгруппируем данные data по категориям целей purpose_category 
# и применим метод agg(), чтобы для каждой группы посчитать sum() и count() 
purpose_grouped = data.groupby('purpose_category').agg({'debt': ['sum', 'count']})
purpose_grouped['conversion'] = purpose_grouped['debt']['sum'] / purpose_grouped['debt']['count']
purpose_grouped['category_part'] = purpose_grouped['debt']['count'] / \
purpose_grouped['debt']['count'].sum()
purpose_grouped

debt        conversion category_part
                          sum  count                         
purpose_category                                             
операции с автомобилем    400   4279   0.093480      0.200600
операции с недвижимостью  780  10751   0.072551      0.504008
получение образования     369   3988   0.092528      0.186958
проведение свадьбы        183   2313   0.079118      0.108434

**Вывод: самая низкая конверсия в категории "операции с недвижимостью" говорит о том, что с большей долей вероятности будет возвращен кредит, если он взят с этой целью, также эта категория составляет наибольшее количество обращений.** 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ: пропуски содержатся в данных о трудовом стаже и о доходах. Возможная причина пропусков - неофициальное трудоустройство и доходы, которые заемщик не может подтвердить заверенной документацией. Также среди заемщиков могут быть дети, родители которых готовы оплачивать кредиты детям, у которых нет ни стажа, ни доходов.* 

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ: среди выборки значений, могут быть значения сильно отличающиеся от основного количества данных и если мы будем заполнять значеними среднего, то выделяемые данные могут повлиять на реальную картину вычислений. Также важно указывать медианные значения именно в каждой категории, потому что в зависимости от категории они могут различаться.* 

### Шаг 4: общий вывод.


Получив таблицу со статистикой о платежеспособности клиентов, прежде чем ее анализировать, 
необходимо было провести предобработку данных.

В первую очередь были обнаружены пропустки в двух столбцах (сумма дохода и стаж работы) 
в большом количестве (из 21525 строк заполнены лишь 19351).
Это значит, что потеряв эти строки, результаты анализа могут быть искажены, 
поэтому было принято решение их заполнить.

На сумму дохода влияет тип занятости, поэтому пропуски были заполнены медианным значением в каждой группе
типа занятости.
По такому же принципу заполнены и медианные значения в столбце с данными о стаже.

Кроме пропусков были обнаружены отрицательные значения, что является аномалией, 
поэтому эти данные были приведены к абсолютным значениям.

В столбце с количеством детей были обнаружены аномальные значения (отрицательное -1 и слишком большое 20).
Строки с этими данными были удалены, чтобы они не исказили итоговые результаты.

Также после приведения к нижнему регистру значений в столбце об образовании, были обнаружены явные дубликаты,
которые были удалены.

После предобработки данных можно было приступать к решению задачи. 
А именно необходимо было выяснить влияет ли семейное положение и количество детей клиента
на факт погашения кредита в срок.

Для принятия решения по этому вопросу были поставлены следующие цели:
    
**1. Выснить, есть ли зависимость между количеством детей и возвратом кредита в срок.**
    
На основании данных о должниках в статистике была рассчитана конверсия в зависомсти от количества детей.
Чем меньше конверсия, тем меньше должников в определенной группе. Так, результаты показали, что самая
низкая конверсия (7.5%)
у семей без детей, в отличие от семей с одним, двумя, тремя или четырьмя детьми
(соответственно 9.2%, 9.4%, 8%, 9.7%).
    
    
**2. Выяснить, есть ли зависимость между семейным положением и возвратом кредита в срок.**
    
Больше половины заемщиков состоят в браке (57%) и конверсия (7.5%) у них ниже, чем у тех категорий граждан, 
которые не состоят в официальном браке (не женат / не замужем (9.7%), гражданский брак(9.3%)), 
самая низкая конверсия у тех, кто в разводе (7%), а также у вдовцов / вдов (6.6%), 
но их суммарно всего 10% от общего количества заемщиков, 
поэтому я бы рекомендовала банкам обращать внимание на семейное положение заемщиков.
    
**3. Выяснить, есть ли зависимость между уровнем дохода и возвратом кредита в срок.**
    
Предварительно все данные о доходах были разделены на четыре категории, чтобы в каждом сегменте было
примерно одинаковое количество заемщиков:
- до 100тыс. руб;
- от 100 тыс. руб. до 150 тыс. руб.;
- от 150 тыс. руб. до 240 тыс. руб.;
- выше 240 тыс. руб.
    
Данные показали, чем выше доходы у заемщика, тем большая вероятность возврата кредита в срок,
т.к. конверсия для категории заемщиков с наибольшими доходами наименьшая (6.9%),
а, значит, и количество должников в этой группе меньше. 
Конверсия у трех других групп в порядке уменьшения доходов следующая: 8.1%, 8.7%, 7.9%.
        
**4. Выяснить, как разные цели кредита влияют на его возврат в срок.**
    
Все цели предварительно были разделены на четыре группы и определены доли этих целей
от общего числа запросов:
- операции с автомобилем (20%);
- операции с недвижимостью (50%);
- получение образования (19%);
- проведение свадьбы (11%).

Cамая низкая конверсия в категории "операции с недвижимостью" (7.2%) говорит о том,
что с большей долей вероятности будет возвращен кредит, если он взят с этой целью,
также эта категория составляет наибольшее количество обращений. 
Конверсия на операции с автомобилем составляет 9.3%, на получение образования 9.2%, 
на проведение свадьбы 7.9%.
    

**Возвращаясь к главному вопросу, 
влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок,
мой вывод следующий: оба эти критерия существенно влияют причем таким образом, что 
идеальный заемщик должен состоять в официальном браке и при этом не иметь детей, кроме этого должен иметь
высокий доход и цель его кредита связана с операциями с недвижимостью. В этом случае у банка будет самый низкий 
процент должников.**
